### ISSUE TITLE
Support for Gamma sampler in JAX

### ISSUE THREAD

**Issue by fehiepsi** [google/jax#552]
Gamma sampler is one of the most popular samplers. Many popular distributions can be derived from Gamma sampler (e.g. Beta, Dirichlet, StudentT). This thread is created to discuss about details for an implementation in JAX.

### Implement Gamma sampler

I made an initial attempt in https://github.com/google/jax/pull/551. Currently, the implementation is a bit slow but I believe that with some help from JAX devs, it will be fast.

TODO:
- [ ] Explore warp-based parallelism as suggested by @slinderman

### Implement JVP rule for shape parameter `a` (or alpha).

There are two recent papers which addresses this:

+ Ref 1. [Implicit Reparameterization Gradients](https://arxiv.org/abs/1805.08498), which is implemented in tensorflow.
+ Ref 2. [Pathwise Derivatives Beyond the Reparameterization Trick](https://arxiv.org/abs/1806.01851), which is implemented in pytorch.

Two references come up with the same formula for gradient of z ~ Gamma(a) w.r.t. a:
```
dz/da = - dCDF(z; a)/da / pdf(z; a)
```
Further derivation leads to:
```
# dz = - dCDF(z; a) / pdf(z; a)
# pdf = z^(a-1) * e^(-z) / Gamma(a)
# CDF(z; a) = IncompleteGamma(a, z) / Gamma(a)
# dCDF(z; a) = (dIncompleteGamma - IncompleteGamma * Digamma(a)) / Gamma(a)
# dz = (IncompleteGamma(a, z) * Digamma(a) - dIncompleteGamma) * e^z / z^(a-1)
```
So to compute `unnormalized_dCDF`, a derivative for IncompleteGamma function is required. [1] and [2] provide different ways to compute this derivative.

Two questions come to my mind:
+ If we have implemented reparameterized version of Gamma, do we need to compute pathwise derivative for all distributions Beta, Dirichlet, StudentT too? In [2], the authors computed pathwise derivative for Gamma and Beta/Dirichlet separately, and did not deal with StudentT. In [1], the authors only provide methods to compute pathwise derivative for Gamma.
+ Is it true (as reported in [1]) that the algorithm in [1] is faster than the algorithm in [2]? Or the performance difference comes from using different frameworks? It seems to me that the authors in [2] only uses first 5 terms in the IncompleteGamma series (and correct the accuracy on extreme cases using polynomials) while the authors of [1] use up to 200 terms. So I feel that the method in [2] would be faster than in [1], but might be less accuracy.

cc @neerajprad

**Comment by brianwa84** [comment]
cc @mfigurnov

**Comment by mattjj** [comment]
@fehiepsi IIUC I think you've fixed this now. I'm going to close this issue, but let me know if I'm mistaken!

**Comment by danleonte** [comment]
hi
is there any support or workaround for the reparametrization trick in the case of truncated distributions? I am particularly interested in gamma, inverse Gaussian and Normal-inverse Gaussian distribution. thanks!

**Comment by brianwa84** [comment]
TFP has a jax variant you could use for these, `from
tensorflow_probability.substrates import jax as tfp`, then go with
tfp.distributions.* You could use a custom distribution like on
tensorflow/probability#1135 (write `tf = tfp.tf2jax`)

Brian Patton | Software Engineer | bjp@google.com

**Comment by srvasude** [comment]
Hi, FWIW `InverseGaussian` has recently be reparameterized in TFP. `NormalInverseGaussian` will also make it there soon (which uses the `InverseGaussian` sampler and hence will be reparameterized).

**Comment by danleonte** [comment]
@srvasude  @brianwa84  thank you very much for your answers. 

please correct me if i got it wrong: your suggestion is to essentially sample from the non-truncated distributions and then only use the samples that are within the truncation bounds?

**Comment by srvasude** [comment]
Sorry, my comment here was just availability of the non-truncated distributions. 

Brian's suggestion contains a way to do generic truncation, but it does the naive thing of sampling and checking if it is in the truncation bounds (which can be sample inefficient).

**Comment by danleonte** [comment]
@srvasude  thanks again for clarifying. 

I understand that the Gamma, InverseGaussian and NormalInverseGaussian are reparametrised by passing uniform samples on [0,1] through the quantile function F^-1.  Assume that we truncate to [low,high]  and let F be the cdf pf the given  (non-truncated) distribution.  instead of sampling and rejecting, wouldn't it be more effective to sample u uniformly on [0,1] and return  F^-1 (y * (F(b)-F(a) + F(a))  ?

**Comment by srvasude** [comment]
Currently these distributions are sampled via rejection reparameterized via the reparameterization trick: https://arxiv.org/abs/1805.08498 (which in essence means overriding the gradient of the samples with gradients of the quantile).

In principle one can use inverse CDF sampling for these truncated distributions and this will work well in some parameter regimes. The issue tends to be that the CDFs / Quantiles for these distributions tend to become ill-behaved in some parameter regimes. For instance, in the TruncatedNormal case if the bounds are really far away from the mean, you'll can get bad samples due to the 'non-invertibility' of these functions far enough out.

Generally, I think one would need to come up with hand tailored rejection samplers (or other methods) for the distributions in question in order to get performance for sampling (in terms of speed and sample quality).

**Comment by brianwa84** [comment]
Actually the issue discussion I pointed out contains a Truncated
distribution which is a quantile transformed Uniform, not a rejection
sampler. In many cases a rejection sampler would yield better samples
(numerically).

**Comment by danleonte** [comment]
> Hi, FWIW `InverseGaussian` has recently be reparameterized in TFP. `NormalInverseGaussian` will also make it there soon (which uses the `InverseGaussian` sampler and hence will be reparameterized).

is there an expected time for this (weeks, months)? thanks a lot for your help!

**Comment by srvasude** [comment]
The distribution is in TFP: https://www.tensorflow.org/probability/api_docs/python/tfp/distributions/NormalInverseGaussian

**Comment by fehiepsi** [comment]
@srvasude Do you know if there is a gamma sampler on XLA, so we can do something similar to the PR that you made for gamma grad? In https://github.com/pyro-ppl/numpyro/issues/955, I saw that TFP gamma sampler is a few times faster than the jax version here.
